### ISSUE TITLE
java.util.NoSuchElementException: key not found: null in optimizing

### ISSUE THREAD

**Issue by andreaTP** [scala-js/scala-js#3053]
while compiling Akka.js bindings I got a NoSuchElementException
A repro could be found here: https://github.com/akka-js/akka.js_bindings/tree/bugInCompiling

both `fast` and `full` are failing with the very same error

**Comment by andreaTP** [comment]
I cannot make a minimal repro but  the bug resides in https://github.com/akka-js/akka.js_bindings/blob/bugInCompiling/bindings.scala#L134 as far as I can understand ...

**Comment by sjrd** [comment]
Thanks, I will have a look tomorrow.

**Comment by sjrd** [comment]
If I enable IR checking in the project, with
```scala
.settings(
  ...
  scalaJSOptimizerOptions ~= { _.withCheckScalaJSIR(true) }
)
```
then I get 13 IR checking errors coming from Akka.js:
```scala
[error] .../akka.js/akka-js-actor-ir-patches/src/main/scala/ActorIrPatch.scala(10:16:Select): Cannot select context$1 of non-class Lakka_actor_Actor
[error] .../akka.js/akka-js-actor-ir-patches/src/main/scala/ActorIrPatch.scala(11:22:Select): Cannot select self$1 of non-class Lakka_actor_Actor
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(134:45:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(139:47:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(149:69:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(157:57:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(115:49:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Children.scala(215:42:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Children) typed as ClassType(Lakka_actor_ActorCell)
[error] .../akka.js/akka-js-actor/js/src/main/scala/akka/dispatch/Mailbox.scala(462:13:VarDef): ClassType(Lakka_dispatch_DefaultSystemMessageQueue) expected but ClassType(Lakka_dispatch_Mailbox) found for tree of type org.scalajs.core.ir.Trees$AsInstanceOf
[error] .../akka.js/akka-js-actor/js/src/main/scala/akka/dispatch/Mailbox.scala(471:9:VarRef): Variable _$this of type ClassType(Lakka_dispatch_DefaultSystemMessageQueue) typed as ClassType(Lakka_dispatch_Mailbox)
[error] .../akka.js/akka-js-actor/js/src/main/scala/akka/dispatch/Mailbox.scala(477:13:VarDef): ClassType(Lakka_dispatch_DefaultSystemMessageQueue) expected but ClassType(Lakka_dispatch_Mailbox) found for tree of type org.scalajs.core.ir.Trees$AsInstanceOf
[error] .../akka.js/akka-js-actor/js/src/main/scala/akka/dispatch/Mailbox.scala(481:10:VarRef): Variable _$this of type ClassType(Lakka_dispatch_DefaultSystemMessageQueue) typed as ClassType(Lakka_dispatch_Mailbox)
[error] .../akka.js/akka-js-actor/shared/src/main/scala/akka/actor/dungeon/Dispatch.scala(31:111:VarRef): Variable _$this of type ClassType(Lakka_actor_dungeon_Dispatch) typed as ClassType(Lakka_actor_ActorCell)
```
It seems your IR patcher is doing something wrong. Since you're playing with an IR patcher, you really should enable IR checking in the Akka.js project to begin with. You're leaking invalid IR in your published artifacts, which can burn anyone at any time.

It might not be the cause of the current issue, but it could be. I will keep investigating a bit, but if it comes down to the published IR being broken, there's nothing we can do (or should do).

**Comment by sjrd** [comment]
The above was on `master` on the mentioned repo. On the `bugInCompiling` branch, I get two *additional* iR checking errors, which seems to be our fault:
```
[error] /akkajs-bindings/bindings.scala(56:7:Closure): Closure parameter Ident(arg$rest,Some(arg$rest)) cannot be a rest parameter
[error] /akkajs-bindings/bindings.scala(56:7:JSSuperBracketCall): JS class type expected but ClassType(Lorg_akkajs_bindings_Actor$$anon$1) found
```

**Comment by andreaTP** [comment]
thanks a lot for the IR checking tip, I will do that, but from a first look they looks all "safe" typecast that simply cannot be checked properly by JSIR, but I will take a second look later of course!

**Comment by andreaTP** [comment]
specifically:
first two are there to be consistent on instances of that `trait` and not on the `trait` itself and `akka-js-actor-ir-patches` is not shipped,
everything regarding `Children` is ok since there is this in akka sources `trait Children { this: ActorCell`
on `Mailbox` of course `DefaultSystemMessageQueue` is a subclass of `Mailbox` itself
on `Dispatch` again is safe since `trait Dispatch { this: ActorCell ⇒` in Akka sources

I do not know how I can turn these "errors" of ... any hint?

**Comment by sjrd** [comment]
> from a first look they looks all "safe" typecast that simply cannot be checked properly by JSIR

There is no such thing as a "safe" typecast that cannot be IR-checked. An IR checker error is like a `VerifyError` on the JVM. It means the bytecode (IR) is *invalid*. If there is a safe typecast that would fix the IR checker error, you have to add it explicitly as an `AsInstanceOf` node in the IR.

**Comment by sjrd** [comment]
> akka-js-actor-ir-patches is not shipped

The fact that I see IR checker errors in that part of the code, in an independent codebase, is proof that is shipped, and *reachable* to boot.

**Comment by sjrd** [comment]
OK the bug is that an *anonymous* JS class that needs to generate run-time overloading dispatch with several possible argument counts (i.e., that needs a `...rest` parameter in the `MethodDef`) creates broken IR, because it creates a `Closure` with that `...rest` parameter, and that is not allowed.

Workaround: rewrite `newAR` to used a *named* JS class rather than an anonymous class:
```scala
  private def newAR(_ar: akkaactor.ActorRef) = {
    @ScalaJSDefined
    class MyActorRef extends ActorRef(_ar) {
      override def tell(msg: Any) = actorRef.tell(msg, ar)
    }
    new MyActorRef
  }
```

**Comment by sjrd** [comment]
Dependency-free minimization:
```scala
@ScalaJSDefined
class Foo extends js.Object {
  def bar(msg: Any): Unit = this.bar(msg, 0)

  def bar(msg: Any, sender: Int): Unit =
    println("bar(" + msg + ", " + sender + ")")
}

object HelloWorld extends js.JSApp {
  def main(): Unit = {
    val ar = new Foo {
      override def bar(msg: Any): Unit = this.bar(msg, 1)
    }
    ar.bar("foobar")
  }
}
```
Which, in turn, can be decomposed in *two independent issues*:

First, the `Closure` with `...rest` problem:
```scala
@ScalaJSDefined
trait Bar extends js.Object {
  def babar(msg: String): Unit
  def babar(msg: String, sender: Int): Unit
}

object HelloWorld extends js.JSApp {
  def main(): Unit = {
    val bar = new Bar {
      def babar(msg: String): Unit = println(s"babar($msg)")
      def babar(msg: String, sender: Int): Unit = println(s"babar($msg, $sender)")
    }
    bar.babar("foobabar")
  }
}
```
and then, another problem that calling `super` from an anonymous JS class also produces invalid IR (invalid `JSSuperBracketCall` node):
```scala
@ScalaJSDefined
class Foo extends js.Object {
  def bar(msg: String): Unit = println(msg)
}

object HelloWorld extends js.JSApp {
  def main(): Unit = {
    val foo = new Foo {
      override def bar(msg: String): Unit = super.bar("foo: " + msg)
    }
    foo.bar("foobar")
  }
}
```

**Comment by sjrd** [comment]
Extracted in two separate issues:

* [ ] #3054 for the overloading issue
* [ ] #3055 for the super call issue

**Comment by gzm0** [comment]
Hahahaha, I guess I'm to blame for these :)

**Comment by andreaTP** [comment]
ok, got it, I will address the IR checks asap, and also I will try to avoid the copy `pos` from one IR to the other.
Really useful as always @sjrd thanks.

**Comment by sjrd** [comment]
Fixed in 3442701679ea5b29f9d2aabb3cc84af8dc8a7221 + ec1489927e47a2e4701bea68fbf21052aefc59ac
